File: apps/firefly/product/client/imp/eos/app/src/main/kotlin/com/miso/noobtest/SignInView.kt
package com.miso.noobtest

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

enum class SignInState {
    ENTER_EMAIL,
    ENTER_CODE
}

@Composable
fun SignInView(
    onAuthenticated: (Boolean) -> Unit // Callback with isNewUser flag
) {
    val serverURL = "http://185.96.221.52:8080"
    var currentState by remember { mutableStateOf(SignInState.ENTER_EMAIL) }
    var email by remember { mutableStateOf("") }
    var code by remember { mutableStateOf("") }
    var isLoading by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf("") }
    val scope = rememberCoroutineScope()

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color(0xFF40E0D0)),
        contentAlignment = Alignment.Center
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(32.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(30.dp)
        ) {
            // Logo
            Text(
                text = "ᕦ(ツ)ᕤ",
                fontSize = 64.sp,
                color = Color.Black
            )

            // Title
            Text(
                text = "Welcome to Firefly",
                fontSize = 28.sp,
                fontWeight = FontWeight.Bold,
                color = Color.Black
            )

            // Content based on state
            when (currentState) {
                SignInState.ENTER_EMAIL -> {
                    EmailEntryView(
                        email = email,
                        onEmailChange = { email = it },
                        isLoading = isLoading,
                        onSendCode = {
                            scope.launch {
                                errorMessage = ""
                                isLoading = true
                                val result = sendCode(serverURL, email)
                                isLoading = false
                                if (result.first) {
                                    currentState = SignInState.ENTER_CODE
                                } else {
                                    errorMessage = result.second
                                }
                            }
                        }
                    )
                }
                SignInState.ENTER_CODE -> {
                    CodeEntryView(
                        email = email,
                        code = code,
                        onCodeChange = { if (it.length <= 4) code = it },
                        isLoading = isLoading,
                        onVerifyCode = {
                            scope.launch {
                                errorMessage = ""
                                isLoading = true
                                val result = verifyCode(serverURL, email, code)
                                isLoading = false
                                if (result.first != null) {
                                    val isNewUser = result.first!!
                                    Storage.saveLoginState(email, true)
                                    Logger.log("[SignIn] User authenticated: $email (new_user: $isNewUser)")
                                    onAuthenticated(isNewUser)
                                } else {
                                    errorMessage = result.second
                                }
                            }
                        },
                        onBackToEmail = {
                            currentState = SignInState.ENTER_EMAIL
                            code = ""
                            errorMessage = ""
                        }
                    )
                }
            }

            // Error message
            if (errorMessage.isNotEmpty()) {
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(containerColor = Color.White.copy(alpha = 0.8f))
                ) {
                    Text(
                        text = errorMessage,
                        color = Color.Red,
                        modifier = Modifier.padding(16.dp),
                        textAlign = TextAlign.Center
                    )
                }
            }
        }

        // Loading overlay
        if (isLoading) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(Color.Black.copy(alpha = 0.3f)),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator(color = Color.White)
            }
        }
    }
}

@Composable
fun EmailEntryView(
    email: String,
    onEmailChange: (String) -> Unit,
    isLoading: Boolean,
    onSendCode: () -> Unit
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(20.dp)
    ) {
        Text(
            text = "Enter your email address",
            color = Color.Black
        )

        TextField(
            value = email,
            onValueChange = onEmailChange,
            placeholder = { Text("email@example.com") },
            enabled = !isLoading,
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email),
            colors = TextFieldDefaults.colors(
                focusedContainerColor = Color.White,
                unfocusedContainerColor = Color.White
            ),
            modifier = Modifier.fillMaxWidth()
        )

        Button(
            onClick = onSendCode,
            enabled = !isLoading && email.isNotEmpty(),
            colors = ButtonDefaults.buttonColors(containerColor = Color.Black),
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Send Code", fontWeight = FontWeight.SemiBold)
        }
    }
}

@Composable
fun CodeEntryView(
    email: String,
    code: String,
    onCodeChange: (String) -> Unit,
    isLoading: Boolean,
    onVerifyCode: () -> Unit,
    onBackToEmail: () -> Unit
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(20.dp)
    ) {
        Text(
            text = "Enter the 4-digit code",
            color = Color.Black
        )

        Text(
            text = "sent to $email",
            fontSize = 12.sp,
            color = Color.Black.copy(alpha = 0.7f)
        )

        TextField(
            value = code,
            onValueChange = onCodeChange,
            placeholder = { Text("0000") },
            enabled = !isLoading,
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
            textStyle = LocalTextStyle.current.copy(
                fontSize = 24.sp,
                fontWeight = FontWeight.Bold,
                textAlign = TextAlign.Center
            ),
            colors = TextFieldDefaults.colors(
                focusedContainerColor = Color.White,
                unfocusedContainerColor = Color.White
            ),
            modifier = Modifier.fillMaxWidth()
        )

        Button(
            onClick = onVerifyCode,
            enabled = !isLoading && code.length == 4,
            colors = ButtonDefaults.buttonColors(containerColor = Color.Black),
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Verify Code", fontWeight = FontWeight.SemiBold)
        }

        TextButton(
            onClick = onBackToEmail,
            enabled = !isLoading
        ) {
            Text("Use different email", color = Color.Black)
        }
    }
}

// Network functions
private suspend fun sendCode(serverURL: String, email: String): Pair<Boolean, String> {
    return withContext(Dispatchers.IO) {
        try {
            val url = URL("$serverURL/api/auth/send-code")
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.doOutput = true

            val jsonBody = JSONObject().apply {
                put("email", email.lowercase().trim())
            }

            connection.outputStream.use { it.write(jsonBody.toString().toByteArray()) }

            val responseCode = connection.responseCode
            val response = connection.inputStream.bufferedReader().use { it.readText() }
            connection.disconnect()

            val json = JSONObject(response)
            val status = json.getString("status")

            if (status == "success") {
                Pair(true, "")
            } else {
                Pair(false, json.optString("message", "Failed to send code"))
            }
        } catch (e: Exception) {
            Pair(false, "Connection error: ${e.message}")
        }
    }
}

private suspend fun verifyCode(serverURL: String, email: String, code: String): Pair<Boolean?, String> {
    return withContext(Dispatchers.IO) {
        try {
            val url = URL("$serverURL/api/auth/verify-code")
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.doOutput = true

            val deviceID = Storage.getDeviceID()
            val jsonBody = JSONObject().apply {
                put("email", email.lowercase().trim())
                put("code", code)
                put("device_id", deviceID)
            }

            connection.outputStream.use { it.write(jsonBody.toString().toByteArray()) }

            val responseCode = connection.responseCode
            val response = connection.inputStream.bufferedReader().use { it.readText() }
            connection.disconnect()

            val json = JSONObject(response)
            val status = json.getString("status")

            if (status == "success") {
                val isNewUser = json.optBoolean("is_new_user", false)
                Pair(isNewUser, "")
            } else {
                Pair(null, json.optString("message", "Verification failed"))
            }
        } catch (e: Exception) {
            Pair(null, "Connection error: ${e.message}")
        }
    }
}

File: apps/firefly/product/client/imp/ios/NoobTest/SignInView.swift
import SwiftUI

struct SignInView: View {
    // Server configuration
    let serverURL = "http://185.96.221.52:8080"

    // State management
    enum SignInState {
        case enterEmail
        case enterCode
    }

    @State private var currentState: SignInState = .enterEmail
    @State private var email: String = ""
    @State private var code: String = ""
    @State private var isLoading: Bool = false
    @State private var errorMessage: String = ""
    @Binding var isAuthenticated: Bool
    @Binding var isNewUser: Bool
    @FocusState private var isCodeFieldFocused: Bool

    var body: some View {
        ZStack {
            // Background color
            Color(red: 255/255, green: 178/255, blue: 127/255)
                .ignoresSafeArea()

            VStack(spacing: 30) {
                // Logo
                Text("ᕦ(ツ)ᕤ")
                    .font(.system(size: 40))
                    .foregroundColor(.black)
                    .padding(.top, 30)  // Move down by ~75% of height
                    .padding(.bottom, 40)

                // Title
                Text("welcome to microclub.")
                    .font(.title)
                    .foregroundColor(.black)

                // Content based on state
                if currentState == .enterEmail {
                    emailEntryView
                } else {
                    codeEntryView
                }

                // Error message
                if !errorMessage.isEmpty {
                    Text(errorMessage)
                        .foregroundColor(.red)
                        .padding()
                        .background(Color.white.opacity(0.8))
                        .cornerRadius(8)
                }

                Spacer()
            }
            .padding()

            // Loading overlay
            if isLoading {
                Color.black.opacity(0.3)
                    .ignoresSafeArea()
                ProgressView()
                    .scaleEffect(1.5)
            }
        }
    }

    // Email entry view
    var emailEntryView: some View {
        VStack(spacing: 20) {
            Text("please enter your email address:")
                .foregroundColor(.black)

            TextField("", text: $email, prompt: Text("email@example.com").foregroundStyle(Color(red: 180/255, green: 140/255, blue: 110/255)))
                .textFieldStyle(RoundedBorderTextFieldStyle())
                .autocapitalization(.none)
                .keyboardType(.emailAddress)
                .tint(.white)  // White cursor
                .disabled(isLoading)
                .onAppear {
                    UIAutomationRegistry.shared.registerTextField(id: "signin-email") { text in
                        self.email = text
                    }
                }

            Button(action: sendCode) {
                Text("log in")
                    .fontWeight(.semibold)
                    .foregroundColor(.white)
                    .frame(maxWidth: .infinity)
                    .padding()
                    .background(Color.black)
                    .cornerRadius(10)
            }
            .disabled(isLoading || email.isEmpty)
            .onAppear {
                UIAutomationRegistry.shared.register(id: "signin-login") {
                    self.sendCode()
                }
            }
        }
        .padding()
    }

    // Code entry view
    var codeEntryView: some View {
        VStack(spacing: 20) {
            Text("enter the 4-digit code")
                .foregroundColor(.black)

            Text("sent to \(email)")
                .foregroundColor(.black.opacity(0.7))

            // Four separate digit boxes with hidden text field
            ZStack {
                // Hidden text field that receives input
                TextField("", text: $code)
                    .keyboardType(.numberPad)
                    .focused($isCodeFieldFocused)
                    .opacity(0.001)
                    .onChange(of: code) { newValue in
                        // Limit to 4 digits only
                        let filtered = newValue.filter { $0.isNumber }
                        if filtered.count > 4 {
                            code = String(filtered.prefix(4))
                        } else if filtered != newValue {
                            code = filtered
                        }
                    }
                    .onAppear {
                        UIAutomationRegistry.shared.registerTextField(id: "signin-code") { text in
                            self.code = String(text.prefix(4).filter { $0.isNumber })
                        }
                    }

                // Visual digit boxes
                HStack(spacing: 12) {
                    ForEach(0..<4, id: \.self) { index in
                        let digit = index < code.count ? String(code[code.index(code.startIndex, offsetBy: index)]) : ""
                        Text(digit)
                            .font(.system(size: 28, weight: .bold))
                            .foregroundColor(.black)
                            .frame(width: 50, height: 60)
                            .background(Color(red: 240/255, green: 160/255, blue: 110/255))  // Darker peach
                            .cornerRadius(8)
                            .overlay(
                                RoundedRectangle(cornerRadius: 8)
                                    .stroke(isCodeFieldFocused && index == code.count ? Color.black : Color.black.opacity(0.3), lineWidth: isCodeFieldFocused && index == code.count ? 2 : 1)
                            )
                    }
                }
            }
            .onTapGesture {
                isCodeFieldFocused = true
            }
            .onAppear {
                // Auto-focus when view appears
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                    isCodeFieldFocused = true
                }
            }

            Button(action: verifyCode) {
                Text("verify")
                    .fontWeight(.semibold)
                    .foregroundColor(.white)
                    .frame(maxWidth: .infinity)
                    .padding()
                    .background(Color.black)
                    .cornerRadius(10)
            }
            .disabled(isLoading || code.count != 4)
            .onAppear {
                UIAutomationRegistry.shared.register(id: "signin-verify") {
                    self.verifyCode()
                }
            }

        }
        .padding()
    }

    // Send verification code
    func sendCode() {
        errorMessage = ""
        isLoading = true

        guard let url = URL(string: "\(serverURL)/api/auth/send-code") else {
            errorMessage = "Invalid server URL"
            isLoading = false
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["email": email.lowercased().trimmingCharacters(in: .whitespaces)]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                isLoading = false

                if let error = error {
                    errorMessage = "Connection error: \(error.localizedDescription)"
                    return
                }

                guard let data = data,
                      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let status = json["status"] as? String else {
                    errorMessage = "Invalid response from server"
                    return
                }

                if status == "success" {
                    currentState = .enterCode
                    Logger.shared.info("[SignIn] Code sent to \(email)")
                } else {
                    errorMessage = json["message"] as? String ?? "Failed to send code"
                    Logger.shared.error("[SignIn] Failed to send code: \(errorMessage)")
                }
            }
        }.resume()
    }

    // Verify code
    func verifyCode() {
        errorMessage = ""
        isLoading = true

        guard let url = URL(string: "\(serverURL)/api/auth/verify-code") else {
            errorMessage = "Invalid server URL"
            isLoading = false
            return
        }

        let deviceID = Storage.shared.getDeviceID()

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "email": email.lowercased().trimmingCharacters(in: .whitespaces),
            "code": code,
            "device_id": deviceID
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                isLoading = false

                if let error = error {
                    errorMessage = "Connection error: \(error.localizedDescription)"
                    return
                }

                guard let data = data,
                      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let status = json["status"] as? String else {
                    errorMessage = "Invalid response from server"
                    return
                }

                if status == "success" {
                    // Check if this is a new user
                    let newUser = json["is_new_user"] as? Bool ?? false
                    let userId = json["user_id"] as? Int ?? 0
                    let userName = json["name"] as? String ?? ""

                    // Save login state including name
                    Storage.shared.saveLoginState(email: email, userId: userId, name: userName, isLoggedIn: true)
                    Logger.shared.info("[SignIn] User authenticated: \(email) (ID: \(userId), name: \(userName), new_user: \(newUser))")

                    // Trigger authentication and set new user flag
                    isNewUser = newUser
                    isAuthenticated = true
                } else {
                    errorMessage = json["message"] as? String ?? "Verification failed"
                    Logger.shared.error("[SignIn] Verification failed: \(errorMessage)")
                }
            }
        }.resume()
    }
}
